README.md
# MMM-ScreenLogic

[MagicMirror²](https://github.com/MichMich/MagicMirror) module used to connect to a local Pentair ScreenLogic pool controller system.

## Installation

1. Navigate into your MagicMirror's `modules` folder and execute `git clone https://github.com/parnic/MMM-ScreenLogic.git`.
2. `cd MMM-ScreenLogic`
3. Execute `npm install --production` to install the node dependencies.
4. Add the module inside `config.js` placing it where you prefer.

## Config

|Option|Type|Description|Default|
|---|---|---|---|
|`coldTemp`|Integer|Show the temperature colored blue if it's at or below this level for pool/spa (requires option `colored`). This is in whatever scale your system is set to (Fahrenheit/Celsius).|`84`|
|`colored`|Boolean|Whether you'd like colored output or not.|`true`|
|`columns`|Integer|How many columns to use to display the data before starting a new row.|`3`|
|`contentClass`|String|The CSS class used to display content values (beneath the header).|`"light"`|
|`controls`|Array|List of controls to show buttons for. Must also set `showControls` to `true`.<br><br>Each entry in this list is an object with a `type` string property and a `name` string to display.<br><br>Valid `type`s:<br>`"circuit"` - toggle a circuit on or off. Must also have an `id` number property defining the circuit ID to set (see [node-screenlogic](https://github.com/parnic/node-screenlogic) documentation for circuit IDs). `name` is an optional string; if not specified, the name of the equipment in the ScreenLogic system will be used.<br>`"heatmode"` - enable or disable the heater for the pool or spa. Must also have a `body` number property that defines which body to toggle the heater for (`0` is the pool, `1` is the spa). Can optionally have a `heatMode` number property that defines which heat mode to set; if not present, defaults to 3 ("heat pump").<br>`"heatpoint"` - set the heat temperature for the pool or spa. Must also have a `body` number property that defines which body to set the heat point for (`0` is the pool, `1` is the spa)|`[]`|
|`hotTemp`|Integer|Show the temperature colored red if it's at or above this level for pool/spa (requires option `colored`). This is in whatever scale your system is set to (Fahrenheit/Celsius).|`90`|
|`serverAddress`|String|The IPv4 address of a ScreenLogic unit to connect to. If not set, the system will search for a unit to connect to. If set, `serverPort` must also be set.| |
|`serverPort`|Integer|The port of a ScreenLogic unit to connect to (usually 80). If not set, the system will search for a unit to connect to. If set, `serverAddress` must also be set.| |
|`showControls`|Boolean|Whether you'd like to show buttons for controlling pool equipment. Must also setup the `controls` array.|`false`|
|`showFreezeMode`|Boolean|Whether you'd like to show a banner when the pool is in freeze mode or not. [added in v1.0.1]|`true`|
|`showOrp`|Boolean|Whether you'd like to show ORP level or not.|`true`|
|`showPH`|Boolean|Whether you'd like to show pH level or not.|`true`|
|`showPHTankLevel`|Boolean|Whether you'd like to show how much pH balancer is in the tank or not. Only functions if `showPH` is also on.|`true`|
|`pHTankLevelMax`|Boolean|If `showPHTankLevel` is enabled, this is the maximum value that the system returns for a full tank. My systems has this always set to 6, but maybe it differs based on what type of pH balancer you're using.|`6`|
|`showPoolTemp`|Boolean|Whether you'd like to show pool temperature or not.|`true`|
|`showSaltLevel`|Boolean|Whether you'd like to show salt level (in PPM) or not.|`true`|
|`showSaturation`|Boolean|Whether you'd like to show saturation/balance or not.|`true`|
|`showSpaTemp`|Boolean|Whether you'd like to show spa temperature or not.|`true`|

Here is an example of an entry in config.js

```js
{
    module: 'MMM-ScreenLogic',
    header: 'Pool info',
    position: 'top_left',
    config: {
        showSpaTemp: false,
        columns: 2,
        contentClass: 'thin',
        showControls: true,
        controls: [
            {type: 'circuit', id: 500},
            {type: 'circuit', id: 505, name: 'Pool'},
            {type: 'heatmode', body: 0, name: 'Pool heater'},
            {type: 'heatpoint', body: 0, name: 'Pool'},
            {type: 'heatmode', body: 1, heatMode: 2, name: 'Spa heater'},
        ]
    }
},
```

## Screenshot

### With color

![Screenshot with color](/screenshot.png?raw=true "colored: true")

## Notes

Pull requests are very welcome! If you'd like to see any additional functionality, don't hesitate to let me know.

This module only works with ScreenLogic controllers on the local network via either a UDP broadcast on 255.255.255.255 or a direct connection if you've specified an address and port in the configuration.

The data is updated when the pool equipment sends an update (which typically happens 0-10 seconds after anything changes), and direct updates are requested after any control is toggled/changed.

When toggling a circuit or changing heat mode, sometimes other circuits are affected. For example, some pools share the same pump for the pool and spa, so when the pool is toggled on the spa must be toggled off. Unfortunately the ScreenLogic system doesn't update its internal status at any predictable rate, so the data on the screen can be wrong immediately after toggling a circuit until the next periodic update runs. If you know of a reliable way around this, please open a pull request!

## Libraries

This uses a Node.JS library I created for interfacing with ScreenLogic controllers over the network: [node-screenlogic](https://github.com/parnic/node-screenlogic), so feel free to check that out for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4.0] - 2022-06-06

### Added

- Added 'reconnecting' overlay message when the connection to the pool equipment was previously established, but is now trying to be re-established.

### Changed

- Added retries for the unit finder so that a temporary loss of connection does not cause the module to get stuck with outdated, non-updating pool data.
- Improved initial "Loading" message to make it clear that it belongs to the ScreenLogic module.
- Reduced load on the pool equipment for the periodic keep-alive messages that the module sends by using a Ping message instead of requesting a full Status update.

## [1.3.0] - 2022-05-24

### Changed

- Switched to a push-based model where the module is told immediately when something in the pool equipment changes rather than polling for an update every 30 minutes.
- Improved support for multiple clients connecting to the same Magic Mirror server. Additional clients will no longer put additional strain on the pool equipment by triggering more connections and requests.
- Changed thresholds for causing the pH tank level to show low and medium colors.

### Fixed

- Fixed a few different ways that the module could stop requesting or processing updates from the pool equipment. It should be much more resilient against failure now.

## [1.2.0] - 2021-10-13

### Added

- Can optionally display pH tank level alongside raw pH value now (on by default).

### Fixed

- Better font size scaling with newer MagicMirror versions.

## [1.1.2] - 2020-04-08

### Fixed

- Fixed which property was being read to determine whether heat mode was enabled or not. Previously the state of the heater itself was being used to display on/off status, now the requested mode is used instead (the heater toggles on/off as part of normal operation even while heating is enabled).

## [1.1.1] - 2020-04-07

### Added

- Ability to specify which heat mode to use when enabling heating for a specific body. Previously mode 1 was always sent which means "solar" while most people probably want mode 3 which is "heat pump".

## [1.1.0] - 2020-04-01

### Added

- Ability to show buttons for controlling pool equipment (with a touch screen, for example). NOTE: running `npm install` again is necessary after upgrading to this version if the heat controls are used.

## [1.0.3] - 2018-04-27

### Changed

- Packaged a newer `node-screenlogic` dependency to fix server broadcast in certain environments.

## [1.0.2] - 2018-04-26

### Added

- New option `updateInterval` to control how often the pool data is updated (default 30 minutes).

## [1.0.1] - 2018-04-25

### Added

- New option `showFreezeMode` shows a banner acrosss the top if the pool is currently in freeze-protection mode.

## [1.0.0] - 2018-04-24

- Initial release.


node_helper.js
var NodeHelper = require('node_helper');

module.exports = NodeHelper.create({
    setCircuit: function(circuitState) {
        setCircuitState(circuitState, (poolStatus) => {
            this.sendSocketNotification('SCREENLOGIC_CIRCUIT_DONE', {circuitState: circuitState, status: poolStatus});
        });
    },

    setHeatpoint: function(heatpoint) {
        setHeatpointState(heatpoint, (poolStatus) => {
            this.sendSocketNotification('SCREENLOGIC_HEATPOINT_DONE', {heatpoint: heatpoint, status: poolStatus});
        });
    },

    setHeatstate: function(heatstate) {
        setHeatstateState(heatstate, (poolStatus) => {
            this.sendSocketNotification('SCREENLOGIC_HEATSTATE_DONE', {heatstate: heatstate, status: poolStatus});
        });
    },

    setLightcmd: function(lightCmd) {
        setLights(lightCmd, (poolStatus) => {
            this.sendSocketNotification('SCREENLOGIC_LIGHTCMD_DONE', {lightCmd: lightCmd, status: poolStatus});
        });
    },

    notifyReconnecting: function() {
        this.sendSocketNotification('SCREENLOGIC_RECONNECTING');
    },

    socketNotificationReceived: function(notification, payload) {
        if (notification === 'SCREENLOGIC_CONFIG') {
            if (!this.config) {
                this.config = payload;
                connect((status) => {
                    this.sendSocketNotification('SCREENLOGIC_RESULT', status);
                }, () => {
                    this.notifyReconnecting();
                });
            } else if (poolData.status) {
                this.sendSocketNotification('SCREENLOGIC_RESULT', poolData);
            }
            // if we don't have a status yet, assume the initial connection is still in progress and this socket notification will be delivered when setup is done
        }
        if (notification === 'SCREENLOGIC_CIRCUIT') {
            this.setCircuit(payload);
        }
        if (notification === 'SCREENLOGIC_HEATPOINT') {
            this.setHeatpoint(payload);
        }
        if (notification === 'SCREENLOGIC_HEATSTATE') {
            this.setHeatstate(payload);
        }
        if (notification === 'SCREENLOGIC_LIGHTCMD') {
            this.setLightcmd(payload);
        }
    }
});

const ScreenLogic = require('node-screenlogic');
const Log = require('logger');
const reconnectDelayMs = 10 * 1000;
const unitFinderTimeoutMs = 5 * 1000;
var foundUnit;
var poolData = {};
var refreshTimer;
var unitFinderRetry;
var unitReconnectTimer;

function connect(cb, reconnectCb) {
    if (!foundUnit && typeof config !== 'undefined' && config.serverAddress && config.serverPort) {
        Log.info(`[MMM-ScreenLogic] connecting directly to configured unit at ${config.serverAddress}:${config.serverPort}`);
        foundUnit = new ScreenLogic.UnitConnection(config.serverPort, config.serverAddress);
    }

    if (foundUnit) {
        setupUnit(cb, reconnectCb);
    } else {
        findServer(cb, reconnectCb);
    }
}

function findServer(cb, reconnectCb) {
    Log.info('[MMM-ScreenLogic] starting search for local units');
    var finder = new ScreenLogic.FindUnits();
    finder.on('serverFound', (server) => {
        finder.close();
        Log.info(`[MMM-ScreenLogic] local unit found at ${server.address}:${server.port}`);

        foundUnit = new ScreenLogic.UnitConnection(server);
        setupUnit(cb, reconnectCb);

        clearInterval(unitFinderRetry);
        unitFinderRetry = null;
    }).on('error', (e) => {
        Log.error(`[MMM-ScreenLogic] error trying to find a server. scheduling a retry in ${reconnectDelayMs / 1000} seconds`);
        Log.error(e);
        resetFoundUnit();
        setTimeout(() => { findServer(cb); }, reconnectDelayMs);
    });

    finder.search();
    unitFinderRetry = setInterval(() => {
        Log.info(`[MMM-ScreenLogic] didn't find any units within ${unitFinderTimeoutMs / 1000} seconds, trying again...`);
        finder.search();
    }, unitFinderTimeoutMs);
}

function resetFoundUnit() {
    foundUnit = null;
    if (refreshTimer) {
        clearInterval(refreshTimer);
        refreshTimer = null;
    }
    if (unitFinderRetry) {
        clearInterval(unitFinderRetry);
        unitFinderRetry = null;
    }
    if (unitReconnectTimer) {
        clearTimeout(unitReconnectTimer);
        unitReconnectTimer = null;
    }
}

function setupUnit(cb, reconnectCb) {
    Log.info('[MMM-ScreenLogic] initial connection to unit...');

    foundUnit.on('error', (e) => {
        Log.error(`[MMM-ScreenLogic] error in unit connection. restarting the connection process in ${reconnectDelayMs / 1000} seconds`);
        Log.error(e);

        reconnectCb();
        resetFoundUnit();
        unitReconnectTimer = setTimeout(() => { connect(cb, reconnectCb); }, reconnectDelayMs);
    }).on('close', () => {
        Log.error(`[MMM-ScreenLogic] unit connection closed unexpectedly. restarting the connection process in ${reconnectDelayMs / 1000} seconds`);

        reconnectCb();
        resetFoundUnit();
        unitReconnectTimer = setTimeout(() => { connect(cb, reconnectCb); }, reconnectDelayMs);
    }).once('loggedIn', () => {
        Log.info('[MMM-ScreenLogic] logged into unit. getting basic configuration...');
        foundUnit.getControllerConfig();
    }).once('controllerConfig', (config) => {
        Log.info('[MMM-ScreenLogic] configuration received. adding client...');
        poolData.controllerConfig = config;
        poolData.degStr = config.degC ? 'C' : 'F';
        foundUnit.addClient(1234);
    }).once('addClient', () => {
        Log.info('[MMM-ScreenLogic] client added successfully and listening for changes');
        foundUnit.getPoolStatus();
        // connection seems to time out every 10 minutes without some sort of request made
        refreshTimer = setInterval(() => { foundUnit.pingServer(); }, 1 * 60 * 1000);
    }).on('poolStatus', (status) => {
        Log.info('[MMM-ScreenLogic] received pool status update');
        poolData.status = status;
        cb(poolData);
    });

    foundUnit.connect();
}

function setCircuitState(circuitState, cb) {
    if (!foundUnit) {
        cb();
        return;
    }

    Log.info(`[MMM-ScreenLogic] setting circuit ${circuitState.id} to ${circuitState.state}`);
    foundUnit.setCircuitState(0, circuitState.id, circuitState.state);
    foundUnit.getPoolStatus();
}

function setHeatpointState(heatpoint, cb) {
    if (!foundUnit) {
        cb();
        return;
    }

    Log.info(`[MMM-ScreenLogic] setting heatpoint for body ${heatpoint.body} to ${heatpoint.temperature} deg`);
    foundUnit.setSetPoint(0, heatpoint.body, heatpoint.temperature);
    foundUnit.getPoolStatus();
}

function setHeatstateState(heatstate, cb) {
    if (!foundUnit) {
        cb();
        return;
    }

    Log.info(`[MMM-ScreenLogic] setting heat state for body ${heatstate.body} to ${heatstate.state}`);
    foundUnit.setHeatMode(0, heatstate.body, heatstate.state);
    foundUnit.getPoolStatus();
}

function setLights(lightCmd, cb) {
    if (!foundUnit) {
        cb();
        return;
    }

    Log.info(`[MMM-ScreenLogic] sending light command ${lightCmd}`);
    foundUnit.sendLightCommand(0, lightCmd);
    foundUnit.getPoolStatus();
}


MMM-ScreenLogic.js
let poolData = {};
let moduleObj;

Module.register('MMM-ScreenLogic',{
    defaults: {
        showPoolTemp: true,
        showSpaTemp: true,
        showPH: true,
        showOrp: true,
        showSaltLevel: true,
        showSaturation: true,
        showFreezeMode: true,
        showControls: false,
        controls: [],
        colored: true,
        coldTemp: 84,
        hotTemp: 90,
        columns: 3,
        contentClass: 'light',
        showPHTankLevel: true,
        pHTankLevelMax: 6
    },

    start: function() {
        // this isn't a great solution...is there a better one? needed to do stuff with buttons
        moduleObj = this;
        if (this.config.showControls && (!this.config.controls || this.config.controls.length === 0)) {
            Log.warn('Controls are enabled, but no controls are configured. See README for info on setting up controls.');
            this.config.showControls = false;
        }

        this.sendSocketNotification('SCREENLOGIC_CONFIG', this.config);
    },

    getStyles: function() {
        return ['screenlogic.css'];
    },

    getDom: function() {
        if (!poolData.status) {
            let wrapper = document.createElement('div');
            wrapper.innerHTML = 'Loading ScreenLogic...';
            wrapper.className += 'dimmed light small text-center';

            return wrapper;
        } else {
            let outermost = document.createElement('div');
            outermost.classList.add('container');

            let reconnectDiv = document.createElement('div');
            reconnectDiv.classList.add('overlay', 'reconnecting', 'd-none');

            let reconnectLabel = document.createElement('div');
            reconnectLabel.classList.add('margin-auto', 'bg-blur');
            reconnectLabel.innerHTML = 'Reconnecting...';
            reconnectDiv.appendChild(reconnectLabel);

            let table = document.createElement('table');
            table.classList.add('base-content', 'small');
            if (this.config.colored) {
                table.classList.add('colored');
            }

            outermost.appendChild(reconnectDiv);
            outermost.appendChild(table);

            let contents = [];

            if (this.config.showPoolTemp) {
                let className = '';
                if (poolData.status.currentTemp[0] <= this.config.coldTemp) {
                    className += ' cold-temp';
                } else if (poolData.status.currentTemp[0] >= this.config.hotTemp) {
                    className += ' hot-temp';
                }

                contents.push({
                    header: 'Pool temp',
                    data: poolData.status.currentTemp[0] + '&deg;' + (!isPoolActive(poolData.status) ? ' (last)' : ''),
                    class: this.config.contentClass + className
                });
            }
            if (this.config.showSpaTemp) {
                let className = '';
                if (poolData.status.currentTemp[1] <= this.config.coldTemp) {
                    className = ' cold-temp';
                } else if (poolData.status.currentTemp[1] >= this.config.hotTemp) {
                    className = ' hot-temp';
                }

                contents.push({
                    header: 'Spa temp',
                    data: poolData.status.currentTemp[1] + '&deg;' + (!isSpaActive(poolData.status) ? ' (last)' : ''),
                    class: this.config.contentClass + className
                });
            }
            if (this.config.showPH) {
                let dataStr = poolData.status.pH
                if (this.config.showPHTankLevel) {
                    let percent = Math.round(((poolData.status.pHTank - 1) / this.config.pHTankLevelMax) * 100)
                    let cls = ''
                    if (this.config.colored) {
                        if (percent <= 17) {
                            cls = 'progress-bar-danger'
                        } else if (percent <= 33) {
                            cls = 'progress-bar-warning'
                        } else {
                            cls = 'progress-bar-success'
                        }
                    }
                    let progBarDiv = `<div class="progress vertical">
                        <div class="progress-bar ${cls}" role="progressbar" aria-valuenow="${percent}" aria-valuemin="0" aria-valuemax="100" style="width: ${percent}%;">
                        </div>
                    </div>`

                    dataStr = `${dataStr} ${progBarDiv}`
                }

                contents.push({
                    header: 'pH',
                    data: dataStr,
                    class: this.config.contentClass
                });
            }
            if (this.config.showOrp) {
                contents.push({
                    header: 'ORP',
                    data: poolData.status.orp,
                    class: this.config.contentClass
                });
            }
            if (this.config.showSaltLevel) {
                contents.push({
                    header: 'Salt PPM',
                    data: poolData.status.saltPPM,
                    class: this.config.contentClass
                });
            }
            if (this.config.showSaturation) {
                contents.push({
                    header: 'Saturation',
                    data: poolData.status.saturation,
                    class: this.config.contentClass
                });
            }
            if (this.config.showControls) {
                for (let control in this.config.controls) {
                    let controlObj = this.config.controls[control];

                    if (controlObj.type === 'circuit') {
                        let name = controlObj.name;
                        for (let circuit in poolData.controllerConfig.bodyArray) {
                            if (poolData.controllerConfig.bodyArray[circuit].circuitId === controlObj.id) {
                                if (!name) {
                                    name = poolData.controllerConfig.bodyArray[circuit].name;
                                }
                            }
                        }

                        let on = false;
                        for (let circuit in poolData.status.circuitArray) {
                            if (poolData.status.circuitArray[circuit].id === controlObj.id) {
                                on = poolData.status.circuitArray[circuit].state !== 0;
                            }
                        }

                        let cls = '';
                        if (this.config.colored) {
                            cls = on ? 'control-on' : 'control-off';
                        }

                        contents.push({
                            data: '<button id="sl-control-' + controlObj.id + '" class="control ' + cls + '" onclick="setCircuit(this)" data-circuit="' +
								controlObj.id + '" data-state="' + (on ? '1' : '0') + '"><div class="content">' +
								name + '</div></button>',
                            class: this.config.contentClass
                        });
                    } else if (controlObj.type === 'heatpoint') {
                        if (controlObj.body < 0 || controlObj.body > poolData.status.setPoint.length) {
                            Log.warn('Invalid body specified for heatpoint');
                            continue;
                        }

                        let temperature = poolData.status.setPoint[controlObj.body];

                        let dataHtml = '<div class="temperature-container">';
                        dataHtml += '<button id="sl-temp-up-'+controlObj.body+'" class="temperature control-off" onclick="setHeatpoint(this, 1)" data-body="'+controlObj.body+'" data-temperature="'+temperature+'"><div class="content">+</div></button>';
                        dataHtml += '<div class="temperature-label">'+controlObj.name+': '+temperature+'&deg;</div>';
                        dataHtml += '<button id="sl-temp-down-'+controlObj.body+'" class="temperature control-off" onclick="setHeatpoint(this, -1)" data-body="'+controlObj.body+'" data-temperature="'+temperature+'"><div class="content">-</div></button>';

                        contents.push({
                            data: dataHtml,
                            class: this.config.contentClass
                        });
                    } else if (controlObj.type === 'heatmode') {
                        if (controlObj.body < 0 || controlObj.body > poolData.status.heatMode.length) {
                            Log.warn('Invalid body specified for heatmode');
                            continue;
                        }

                        let on = poolData.status.heatMode[controlObj.body] !== 0;
                        let mode = typeof controlObj.heatMode === 'number' ? controlObj.heatMode : 3;

                        let cls = '';
                        if (this.config.colored) {
                            cls = on ? 'control-on' : 'control-off';
                        }

                        contents.push({
                            data: '<button id="sl-heat-' + controlObj.body + '" class="control ' + cls + '" onclick="setHeatmode(this)" data-body="' +
								controlObj.body + '" data-state="' + (on ? '1' : '0') + '" data-mode="' + mode.toString() + '"><div class="content">' +
								controlObj.name + '</div></button>',
                            class: this.config.contentClass
                        });
                    } else {
                        Log.warn('circuit with unknown type, unable to display:');
                        Log.warn(controlObj);
                    }
                }
            }

            let headerRow = null;
            let contentRow = null;

            if (this.config.showFreezeMode && poolData.status.freezeMode !== 0) {
                let row = document.createElement('tr');
                table.appendChild(row);
                row.className = 'cold-temp';
                let cell = document.createElement('th');
                row.appendChild(cell);
                cell.colSpan = this.config.columns;
                cell.innerHTML = '<center>FREEZE MODE</center>';
            }

            let cols = -1;
            for (let item in contents) {
                cols++;
                if (cols % this.config.columns === 0) {
                    headerRow = document.createElement('tr');
                    contentRow = document.createElement('tr');
                    table.appendChild(headerRow);
                    table.appendChild(contentRow);
                }

                if (contents[item].header) {
                    let headerCell = document.createElement('th');
                    headerCell.innerHTML = contents[item].header;
                    headerRow.appendChild(headerCell);
                }

                let contentCell = document.createElement('td');
                contentCell.innerHTML = contents[item].data;
                contentCell.className = contents[item].class;
                contentRow.appendChild(contentCell);
            }

            return outermost;
        }
    },

    socketNotificationReceived: function(notification, payload) {
        if (notification === 'SCREENLOGIC_RESULT') {
            poolData = payload;
            this.updateDom();
            showReconnectOverlay(false);
        } else if (notification === 'SCREENLOGIC_CIRCUIT_DONE'
			|| notification === 'SCREENLOGIC_HEATSTATE_DONE'
			|| notification === 'SCREENLOGIC_HEATPOINT_DONE') {
            poolData.status = payload.status;
            this.updateDom();
            showReconnectOverlay(false);
        } else if (notification === 'SCREENLOGIC_RECONNECTING') {
            showReconnectOverlay(true);
        }
    },
});

function showReconnectOverlay(show) {
    let element = document.querySelector('.MMM-ScreenLogic .reconnecting');
    if (!element || !element.classList) {
        return;
    }

    if (show) {
        element.classList.remove('d-none');
    } else {
        element.classList.add('d-none');
    }
}

const SPA_CIRCUIT_ID = 500;
const POOL_CIRCUIT_ID = 505;

function isPoolActive(status) {
    for (let i = 0; i < status.circuitArray.length; i++) {
        if (status.circuitArray[i].id === POOL_CIRCUIT_ID) {
            return status.circuitArray[i].state === 1;
        }
    }
}

function hasSpa(status) {
    for (let i = 0; i < status.circuitArray.length; i++) {
        if (status.circuitArray[i].id === SPA_CIRCUIT_ID) {
            return true;
        }
    }

    return false;
}

function isSpaActive(status) {
    for (let i = 0; i < status.circuitArray.length; i++) {
        if (status.circuitArray[i].id === SPA_CIRCUIT_ID) {
            return status.circuitArray[i].state === 1;
        }
    }
}

function setCircuit(e) {
    let circuitId = parseInt(e.dataset.circuit);
    let on = e.dataset.state !== '0';
    moduleObj.sendSocketNotification('SCREENLOGIC_CIRCUIT', {id: circuitId, state: on ? 0 : 1});
    e.classList.remove('control-on', 'control-off');
}

function setHeatmode(e) {
    let bodyId = parseInt(e.dataset.body);
    let on = e.dataset.state !== '0';
    let mode = e.dataset.mode;
    moduleObj.sendSocketNotification('SCREENLOGIC_HEATSTATE', {body: bodyId, state: on ? 0 : parseInt(mode)});
    e.classList.remove('control-on', 'control-off');
}

function setHeatpoint(e, tempChange) {
    let bodyId = parseInt(e.dataset.body);
    let temp = parseInt(e.dataset.temperature) + tempChange;
    moduleObj.sendSocketNotification('SCREENLOGIC_HEATPOINT', {body: bodyId, temperature: temp});
    e.classList.remove('control-on', 'control-off');
}
